src/Catalyst.Core.Modules.P2P.Discovery.Hastings.Tests/UnitTests/DiscoveryTestBuilder.cs

```diff
@@ -0,0 +1,327 @@
+
+            _careTaker.Add(memento);
+
+            return this;
+        }
+
+        public DiscoveryTestBuilder WithStepProposal(IHastingsOriginator stateCandidate = default,
+            bool mock = false,
+            PeerId peer = default,
+            INeighbours neighbours = default,
+            ICorrelationId expectedPnr = default)
+        {
+            var pnrCorrelationId = expectedPnr ?? CorrelationId.GenerateCorrelationId();
+
+            _currentState = mock
+                ? stateCandidate ?? DiscoveryHelper.MockOriginator(peer, neighbours)
+                : stateCandidate ?? DiscoveryHelper.SubOriginator(peer, neighbours, pnrCorrelationId);
+
+            return this;
+        }
+
+        public DiscoveryTestBuilder WithCareTaker(IHastingsCareTaker hastingsCareTaker = default,
+            IEnumerable<IHastingsMemento> history = default)
+        {
+            _careTaker = hastingsCareTaker ?? DiscoveryHelper.MockCareTaker(history);
+            return this;
+        }
+
+        public sealed class HastingDiscoveryTest : HastingsDiscovery
+        {
+            private HastingDiscoveryTest(ILogger logger = default,
+                IPeerRepository peerRepository = default,
+                IPeerSettings peerSettings = default,
+                IDns dns = default,
+                IPeerClient peerClient = default,
+                IPeerMessageCorrelationManager peerMessageCorrelationManager = default,
+                ICancellationTokenProvider cancellationTokenProvider = default,
+                IEnumerable<IPeerClientObservable> peerClientObservables = default,
+                bool autoStart = true,
+                int peerDiscoveryBurnIn = 10,
+                IHastingsOriginator state = default,
+                IHastingsCareTaker hastingsCareTaker = default,
+                int millisecondsTimeout = 10_000,
+                int hasValidCandidatesCheckMillisecondsFrequency = 1_000,
+                IScheduler scheduler = null)
+                : base(logger ?? Substitute.For<ILogger>(),
+                    peerRepository ?? Substitute.For<IPeerRepository>(),
+                    dns ?? DiscoveryHelper.MockDnsClient(peerSettings),
+                    peerSettings ?? PeerSettingsHelper.TestPeerSettings(),
+                    peerClient ?? Substitute.For<IPeerClient>(),
+                    peerMessageCorrelationManager ?? DiscoveryHelper.MockCorrelationManager(scheduler),
+                    cancellationTokenProvider ?? new CancellationTokenProvider(),
+                    peerClientObservables,
+                    autoStart,
+                    peerDiscoveryBurnIn,
+                    state,
+                    hastingsCareTaker,
+                    millisecondsTimeout,
+                    hasValidCandidatesCheckMillisecondsFrequency) { }
+
+            internal static HastingDiscoveryTest GetTestInstanceOfDiscovery(ILogger logger,
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Method has 15 parameters, which is greater than the 7 authorized.](https://app.codacy.com/app/catalyst-network/Catalyst.Node/pullRequest?prid=4318201)